File: apps/api/app/services/content/editorial_image_utils.py
"""Editorial hero image helpers: stale check, alt sync, publishing sync."""

from __future__ import annotations

from typing import Any

from app.schemas.content_seo_editorial import (
    EditorialApprovedImageBackup,
    EditorialArticlePayload,
    EditorialImagePayload,
    EditorialPublishingPayload,
    normalize_editorial_image_payload,
)
from app.services.content.editorial_image_filename import (
    build_editorial_image_filename,
    filename_slug_from_image_filename,
)
from app.services.content.editorial_image_storage import (
    is_shopify_image_publishable,
    storage_warning_if_needed as resolve_editorial_storage_warning,
)
from app.utils.slug import slugify

IMAGE_STALE_MESSAGE = (
    "L'immagine potrebbe non essere aggiornata rispetto all'ultima versione dell'articolo."
)
IMAGE_STALE_PUBLISH_WARNING = (
    "L'immagine approvata potrebbe non essere allineata all'ultima versione dell'articolo."
)
NO_APPROVED_IMAGE_WARNING = "Nessuna immagine approvata associata all'articolo."
FILENAME_STALE_MESSAGE = (
    "Il titolo articolo è cambiato: il nome file SEO potrebbe non essere più allineato."
)
DEFAULT_IMAGE_ALT = "Immagine articolo Solmielato"


def _read_str_field(payload: dict[str, Any] | None, *keys: str) -> str | None:
    if not payload or not isinstance(payload, dict):
        return None
    for key in keys:
        value = payload.get(key)
        if value is not None and str(value).strip():
            return str(value).strip()
    return None


def _payload_is_present(payload: dict[str, Any] | EditorialImagePayload | None) -> bool:
    if payload is None:
        return False
    if isinstance(payload, EditorialImagePayload):
        return payload.image_status != "not_generated"
    if not isinstance(payload, dict):
        return False
    status = _read_str_field(payload, "imageStatus", "image_status") or "not_generated"
    return status != "not_generated"


def resolve_editorial_image_alt(
    article: EditorialArticlePayload | dict[str, Any] | None,
    brief: dict[str, Any] | None,
    item_title: str | None,
) -> str:
    if article is not None:
        title = (
            article.title.strip()
            if isinstance(article, EditorialArticlePayload)
            else str(article.get("title") or article.get("Title") or "").strip()
        )
        if title:
            return title
    if brief and isinstance(brief, dict):
        proposed = str(brief.get("proposedTitle") or brief.get("proposed_title") or "").strip()
        if proposed:
            return proposed
    if item_title and item_title.strip():
        return item_title.strip()
    return DEFAULT_IMAGE_ALT


def is_image_stale(
    article_payload: dict[str, Any] | EditorialArticlePayload | None,
    image_payload: dict[str, Any] | EditorialImagePayload | None,
) -> bool:
    if article_payload is None or not _payload_is_present(image_payload):
        return False
    source_hash = _read_str_field(
        image_payload if isinstance(image_payload, dict) else image_payload.model_dump(by_alias=True),
        "sourceArticleHash",
        "source_article_hash",
    )
    if isinstance(image_payload, EditorialImagePayload):
        source_hash = image_payload.source_article_hash or source_hash
    article_hash = _read_str_field(
        article_payload if isinstance(article_payload, dict) else article_payload.model_dump(by_alias=True),
        "articleHash",
        "article_hash",
    )
    if isinstance(article_payload, EditorialArticlePayload):
        article_hash = article_payload.article_hash or article_hash
    if not source_hash or not article_hash:
        return False
    return article_hash != source_hash


def is_image_filename_stale(
    image_payload: EditorialImagePayload,
    article_title: str,
) -> bool:
    if not image_payload.image_filename:
        return False
    current_slug = slugify(article_title.strip())
    file_slug = filename_slug_from_image_filename(image_payload.image_filename)
    if not current_slug or not file_slug:
        return False
    if file_slug.endswith("-v2") or "-v" in file_slug:
        base_file = file_slug.rsplit("-v", 1)[0]
        return base_file != current_slug[: len(base_file)]
    return file_slug != current_slug


def _effective_publishing_image(image_payload: EditorialImagePayload) -> EditorialImagePayload:
    if image_payload.image_status in ("generated", "uploaded") and image_payload.approved_image_backup:
        backup = image_payload.approved_image_backup
        return image_payload.model_copy(
            update={
                "image_status": "approved",
                "image_url": backup.image_url,
                "image_alt": backup.image_alt,
                "image_storage_path": backup.image_storage_path,
                "image_filename": backup.image_filename,
                "shopify_image_ready": backup.shopify_image_ready,
            }
        )
    return image_payload


def sync_image_alt_from_article(
    image_payload: EditorialImagePayload,
    article: EditorialArticlePayload | dict[str, Any],
    *,
    brief: dict[str, Any] | None = None,
    item_title: str | None = None,
) -> EditorialImagePayload:
    alt = resolve_editorial_image_alt(article, brief, item_title)
    return image_payload.model_copy(update={"image_alt": alt})


def sync_approved_image_to_publishing(
    publishing: EditorialPublishingPayload,
    image_payload: EditorialImagePayload,
) -> EditorialPublishingPayload:
    effective = _effective_publishing_image(image_payload)
    if effective.image_status != "approved":
        return publishing
    if not effective.shopify_image_ready or not effective.image_url:
        return publishing
    return publishing.model_copy(
        update={
            "image_url": effective.image_url,
            "image_alt": effective.image_alt or publishing.image_alt,
        }
    )


def compute_shopify_image_ready(image_url: str | None) -> bool:
    return is_shopify_image_publishable(image_url)


def build_approved_image_backup(image_payload: EditorialImagePayload) -> EditorialApprovedImageBackup:
    return EditorialApprovedImageBackup(
        image_url=image_payload.image_url,
        image_storage_path=image_payload.image_storage_path,
        image_filename=image_payload.image_filename,
        image_alt=image_payload.image_alt,
        image_hash=image_payload.image_hash,
        image_approved_at=image_payload.image_approved_at,
        shopify_image_ready=image_payload.shopify_image_ready,
        image_width=image_payload.image_width,
        image_height=image_payload.image_height,
        image_mime_type=image_payload.image_mime_type,
    )


def storage_warning_if_needed(
    shopify_ready: bool,
    *,
    effective_provider: str | None = None,
    shopify_connected: bool | None = None,
    can_upload_files: bool | None = None,
    upload_error: str | None = None,
) -> str | None:
    return resolve_editorial_storage_warning(
        shopify_ready,
        effective_provider=effective_provider,
        shopify_connected=shopify_connected,
        can_upload_files=can_upload_files,
        upload_error=upload_error,
    )


def is_image_publish_sync_stale(
    image_payload: EditorialImagePayload,
    publishing: EditorialPublishingPayload,
) -> bool:
    if image_payload.image_status != "approved":
        return False
    if not image_payload.image_url:
        return False
    if publishing.image_url and publishing.image_url != image_payload.image_url:
        return True
    if publishing.image_alt and image_payload.image_alt and publishing.image_alt != image_payload.image_alt:
        return True
    return False


def is_image_shopify_synced(image_payload: EditorialImagePayload) -> bool:
    if not image_payload.shopify_image_synced_at:
        return False
    if image_payload.image_approved_at:
        return image_payload.shopify_image_synced_at >= image_payload.image_approved_at
    return True


def empty_editorial_image_payload() -> EditorialImagePayload:
    return EditorialImagePayload()


def normalize_image_payload(raw: dict[str, Any] | None) -> EditorialImagePayload:
    if not raw:
        return empty_editorial_image_payload()
    return normalize_editorial_image_payload(raw)


def refresh_image_filename_from_title(
    image_payload: EditorialImagePayload,
    article_title: str,
    *,
    version_hint: str | None = None,
) -> EditorialImagePayload:
    alt = article_title.strip() or DEFAULT_IMAGE_ALT
    filename = build_editorial_image_filename(alt)
    if version_hint:
        from app.services.content.editorial_image_filename import resolve_unique_editorial_image_filename

        filename = resolve_unique_editorial_image_filename(
            alt,
            existing_filenames=set(),
            version_hint=version_hint,
        )
    return image_payload.model_copy(
        update={
            "image_alt": alt,
            "image_filename": filename,
        }
    )

File: apps/web/src/components/content/editorial/editorial-image-utils.ts
import type { EditorialImagePayload, EditorialImageStatus } from "@gcr/shared";

export const IMAGE_STALE_MESSAGE =
  "L'immagine potrebbe non essere aggiornata rispetto all'ultima versione dell'articolo.";

export const IMAGE_STALE_PUBLISH_WARNING =
  "L'immagine approvata potrebbe non essere allineata all'ultima versione dell'articolo.";

export const NO_APPROVED_IMAGE_WARNING = "Nessuna immagine approvata associata all'articolo.";

export const PUBLIC_STORAGE_WARNING =
  "Storage pubblico immagini non configurato: l'immagine non può essere inviata a Shopify.";

export const SHOPIFY_NOT_CONNECTED_WARNING =
  "Shopify non connesso: connetti lo shop per caricare l'immagine su Shopify Files.";

export const SHOPIFY_SCOPE_MISSING_WARNING =
  "Per caricare immagini su Shopify serve il permesso write_files o write_images. Aggiorna gli scope della Custom App Shopify.";

export const SHOPIFY_UPLOAD_FAILED_WARNING =
  "Upload Shopify Files fallito. Usa «Riprova upload su Shopify».";

export const FILENAME_STALE_MESSAGE =
  "Il titolo articolo è cambiato: il nome file SEO potrebbe non essere più allineato.";

export const DEFAULT_IMAGE_PROVIDER_SIZE = "1536x1024";
export const DEFAULT_IMAGE_FINAL_SIZE = "1600x900";
export const IMAGE_POST_PROCESSING_LABEL = "crop 16:9 + resize";

const STATUS_LABELS: Record<EditorialImageStatus, string> = {
  not_generated: "Non generata",
  generated: "Generata",
  uploaded: "Caricata su Shopify",
  upload_error: "Upload fallito",
  approved: "Approvata",
};

export function getImageStatusLabel(status: EditorialImageStatus | undefined): string {
  if (!status) return STATUS_LABELS.not_generated;
  return STATUS_LABELS[status] ?? status;
}

export function formatImageCost(cost: number | null | undefined): string {
  if (cost == null || Number.isNaN(cost)) return "—";
  return `$${cost.toFixed(4)}`;
}

export function formatImageUpdatedAt(value: string | null | undefined): string {
  if (!value) return "—";
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return value;
  return date.toLocaleString("it-IT");
}

export function formatImageDimensions(image: EditorialImagePayload): string {
  if (image.imageFinalSize) {
    return image.imageFinalSize.replace("x", "×");
  }
  if (image.imageWidth && image.imageHeight) {
    return `${image.imageWidth}×${image.imageHeight}`;
  }
  return DEFAULT_IMAGE_FINAL_SIZE.replace("x", "×");
}

export function formatImageProviderSize(image: EditorialImagePayload): string {
  const size = image.imageProviderSize ?? DEFAULT_IMAGE_PROVIDER_SIZE;
  return size.replace("x", "×");
}

export function formatImageFinalSize(image: EditorialImagePayload): string {
  return formatImageDimensions(image);
}

export function emptyEditorialImagePayload(): EditorialImagePayload {
  return { imageStatus: "not_generated", imagePrompt: "" };
}

function normalizeImageStatus(value: unknown): EditorialImageStatus {
  const status = String(value ?? "not_generated");
  if (
    status === "generated" ||
    status === "uploaded" ||
    status === "upload_error" ||
    status === "approved"
  ) {
    return status;
  }
  return "not_generated";
}

export function parseEditorialImagePayload(
  raw: EditorialImagePayload | Record<string, unknown> | null | undefined,
): EditorialImagePayload {
  if (!raw || typeof raw !== "object") return emptyEditorialImagePayload();
  const record = raw as Record<string, unknown>;
  const status = normalizeImageStatus(record.imageStatus ?? record.image_status);
  const imageUrl = (record.imageUrl ?? record.image_url ?? record.imagePublicUrl ?? record.image_public_url ?? null) as
    | string
    | null;
  return {
    imageStatus: status,
    imagePrompt: String(record.imagePrompt ?? record.image_prompt ?? ""),
    imageRevisionNote: (record.imageRevisionNote ?? record.image_revision_note ?? null) as string | null,
    imageModel: (record.imageModel ?? record.image_model ?? null) as string | null,
    imageAlt: (record.imageAlt ?? record.image_alt ?? null) as string | null,
    imageUrl,
    imageStoragePath: (record.imageStoragePath ?? record.image_storage_path ?? null) as string | null,
    imageFilename: (record.imageFilename ?? record.image_filename ?? null) as string | null,
    imageOriginalProviderFilename: (record.imageOriginalProviderFilename ??
      record.image_original_provider_filename ??
      null) as string | null,
    imageWidth: (record.imageWidth ?? record.image_width ?? null) as number | null,
    imageHeight: (record.imageHeight ?? record.image_height ?? null) as number | null,
    imageAspectRatio: (record.imageAspectRatio ?? record.image_aspect_ratio ?? null) as string | null,
    imageMimeType: (record.imageMimeType ?? record.image_mime_type ?? null) as string | null,
    imageFileExtension: (record.imageFileExtension ?? record.image_file_extension ?? null) as string | null,
    imageProviderSize: (record.imageProviderSize ?? record.image_provider_size ?? null) as string | null,
    imageFinalSize: (record.imageFinalSize ?? record.image_final_size ?? null) as string | null,
    imageGenerationCost: (record.imageGenerationCost ?? record.image_generation_cost ?? null) as number | null,
    imageGenerationLogId: (record.imageGenerationLogId ?? record.image_generation_log_id ?? null) as string | null,
    imageApprovedAt: (record.imageApprovedAt ?? record.image_approved_at ?? null) as string | null,
    imageHash: (record.imageHash ?? record.image_hash ?? null) as string | null,
    sourceArticleHash: (record.sourceArticleHash ?? record.source_article_hash ?? null) as string | null,
    accessToken: (record.accessToken ?? record.access_token ?? null) as string | null,
    updatedAt: (record.updatedAt ?? record.updated_at ?? null) as string | null,
    skillPackUsed: String(record.skillPackUsed ?? record.skill_pack_used ?? ""),
    skillPackVersion: String(record.skillPackVersion ?? record.skill_pack_version ?? ""),
    shopifyImageReady: Boolean(record.shopifyImageReady ?? record.shopify_image_ready ?? false),
    imageStorageProvider: (record.imageStorageProvider ?? record.image_storage_provider ?? null) as string | null,
    shopifyFileId: (record.shopifyFileId ?? record.shopify_file_id ?? null) as string | null,
    shopifyMediaGid: (record.shopifyMediaGid ?? record.shopify_media_gid ?? null) as string | null,
    shopifyFileStatus: (record.shopifyFileStatus ?? record.shopify_file_status ?? null) as string | null,
    shopifyUploadedAt: (record.shopifyUploadedAt ?? record.shopify_uploaded_at ?? null) as string | null,
    imageUploadError: (record.imageUploadError ?? record.image_upload_error ?? null) as string | null,
    imagePublicUrl: (record.imagePublicUrl ?? record.image_public_url ?? null) as string | null,
    shopifyImageSyncedAt: (record.shopifyImageSyncedAt ?? record.shopify_image_synced_at ?? null) as string | null,
    shopifyImageAltSynced: (record.shopifyImageAltSynced ?? record.shopify_image_alt_synced ?? null) as string | null,
    shopifyImageFilenameSynced: (record.shopifyImageFilenameSynced ??
      record.shopify_image_filename_synced ??
      null) as string | null,
    approvedImageBackup: (record.approvedImageBackup ?? record.approved_image_backup ?? null) as EditorialImagePayload["approvedImageBackup"],
    aiGeneration: (record.aiGeneration ?? record.ai_generation ?? undefined) as EditorialImagePayload["aiGeneration"],
  };
}

export function hasShopifyCdnUrl(image: EditorialImagePayload): boolean {
  const url = (image.imageUrl ?? image.imagePublicUrl ?? "").trim().toLowerCase();
  if (!url) return false;
  return url.includes("cdn.shopify.com") || url.includes("shopifycdn.com");
}

export function hasGeneratedImage(image: EditorialImagePayload): boolean {
  return (
    image.imageStatus === "generated" ||
    image.imageStatus === "uploaded" ||
    image.imageStatus === "upload_error" ||
    image.imageStatus === "approved"
  );
}

export function canApproveImage(image: EditorialImagePayload): boolean {
  if (image.imageStatus === "upload_error" || image.imageStatus === "approved") {
    return false;
  }
  if (image.imageStatus !== "generated" && image.imageStatus !== "uploaded") {
    return false;
  }
  return Boolean(image.imageUrl && image.shopifyImageReady);
}

export function resolveImageStorageWarning(
  image: EditorialImagePayload,
  options?: { canWriteFiles?: boolean; shopifyConnected?: boolean },
): string | null {
  if (image.shopifyImageReady && hasShopifyCdnUrl(image)) {
    return null;
  }
  if (image.imageStatus === "upload_error") {
    return image.imageUploadError
      ? `${SHOPIFY_UPLOAD_FAILED_WARNING} ${image.imageUploadError}`
      : SHOPIFY_UPLOAD_FAILED_WARNING;
  }
  if (image.imageStorageProvider === "shopify_files" || image.imageStorageProvider === null) {
    if (options?.shopifyConnected === false) {
      return SHOPIFY_NOT_CONNECTED_WARNING;
    }
    if (options?.canWriteFiles === false) {
      return SHOPIFY_SCOPE_MISSING_WARNING;
    }
    if (image.imageStatus === "generated" && !image.shopifyImageReady) {
      return "Elaborazione Shopify in corso o upload non ancora completato.";
    }
  }
  if (!image.shopifyImageReady) {
    return PUBLIC_STORAGE_WARNING;
  }
  return null;
}

export function isImageFilenameStale(image: EditorialImagePayload, articleTitle: string): boolean {
  if (!image.imageFilename || !articleTitle.trim()) return false;
  const slugify = (value: string) =>
    value
      .normalize("NFKD")
      .replace(/[\u0300-\u036f]/g, "")
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-+|-+$/g, "");
  const titleSlug = slugify(articleTitle).slice(0, 90);
  const fileSlug = image.imageFilename.replace(/\.jpe?g$/i, "").toLowerCase();
  const baseFile = fileSlug.replace(/-v\d+$/, "").replace(/-[a-f0-9]{6}$/, "");
  return baseFile !== titleSlug && fileSlug !== titleSlug;
}

export function hasApprovedImageForPublish(image: EditorialImagePayload): boolean {
  if (image.imageStatus === "approved") return true;
  return image.imageStatus === "generated" && Boolean(image.approvedImageBackup);
}
